### ISSUE TITLE
reMarkable 2 wakeup timing

### ISSUE THREAD

**Issue by snelg** [koreader/koreader#7139]
>     * I tried various non-zero values in `scheduleIn`. Values as low as 0.1 worked, so I left a little buffer and set it to 0.5.
Possible to make this value customizable?
I had to increase to 1.5 to make it work reliable for my rm2. btw, I had [this fix](https://github.com/toltec-dev/toltec/pull/183) installed, which was delaying the suspend, I believe.

_Originally posted by @wordpr3ss in https://github.com/koreader/koreader/issues/7065#issuecomment-759622353_

**Comment by snelg** [comment]
Dang, didn't consider pre-suspend scripts that might delay the actual suspend longer than usual
Should probably figure out a cleaner solution than my hacky "wait an arbitrary amount of time, and just assume that it was long enough that suspend completed."

Quick recap: after rm2 suspends, when you press the power button to wake up, it does NOT send a message to any of the `/dev/input/eventN`

I'm not an expert on system events and such. I do know that `dmesg` output includes a few particular strings after the system wakes up. So here's a rough pseudocode sketch of a potential still-hacky solution:
(sorry, the pseudocode is more C-ish than Lua-ish)

```
function watchForWakeup() {
  if ('dmesg' output contains "PM: suspend exit") {
    triggerResumeProcesses
  } else {
    scheduleIn(0.5, watchForWakeup)
  }
}

execute('systemctl suspend')
watchForWakeup()
```

Will someone with more reMarkable experience give me some feedback for cleaner ways to watch for system messages? @Eeems maybe?

**Comment by Frenzie** [comment]
As far as wait an arbitrary time goes, it'd probably make more sense to just make it higher rather than customizable.

**Comment by pazos** [comment]
Probably the fix would be out of date when WiFi support is merged. And then your original value would work for all rm2 users.

Also, it seems there're quite a few different patches applied by toltec, oxide and the other Rm hacks available. Including backports, fixes and other workarounds for different problems that might not be known here.

I would suggest to keep the barebones implementation working (ie: KOReader + ddvk's framebuffer) and each launcher/repo/thing that wants to package the app and finds something that doesn't work on barebones open a new issue so we can fix it here.

Otherwise it would be a bit nightmare-ish to keep follow of all changes done on each project and these kind of bugs would be very difficult to repro.

**Comment by snelg** [comment]
> As far as wait an arbitrary time goes, it'd probably make more sense to just make it higher rather than customizable.

I concur, *if* it has to be arbitrary. The annoying thing is that there's no visual response until the arbitrary timer expires. So while it would certainly be "safe" to set the timer for, say, 10 seconds, it would be frustrating to always have to wait almost 10 seconds after pressing the power button before the device has any visual response.

**Comment by Eeems** [comment]
You could probably hook into this: https://github.com/toltec-dev/toltec/blob/testing/package/rm2-suspend-fix/rm2-suspend-fix.sh

**Comment by pazos** [comment]
There's no easy way to do that. If the platform is open enough to have different "callers" you might ends with a complete way of dealing with suspend, unexpected and expected wake events, like in [kobo](https://github.com/koreader/koreader/blob/master/frontend/device/kobo/device.lua#L539-L738). If the platform is closed enough that you're pretty sure you're the only launcher available you can simplify all of this by timming the device and setting a conservative value, like in [cervantes](https://github.com/koreader/koreader/blob/master/platform/cervantes/suspend.sh)

Still think that the community around the ReMarkable is the best for ReMarkable users, but might hurt the ability for us to give support if each workaround is implemented outside this very repo.

**Comment by Eeems** [comment]
> 
> Still think that the community around the ReMarkable is the best for ReMarkable users, but might hurt the ability for us to give support if each workaround is implemented outside this very repo.

There are plans in the work to handle displaying the suspend screen outside of launchers/applications to a separate service. As for handling resume, that would be up to the launcher/application still. Depending on the launcher it either handles it or expects the application to handle it. For KOReader though it would likely make sense to have an optional suspend script like the one I linked to poke KOReader on resume, and add some mechanism to disable it that launchers can handle setting (Probably an empty file in `/tmp`).

**Comment by snelg** [comment]
Yeah, one of my many shortcomings in these discussions is that I *only* care about KOReader and have no plans to run any other apps on my rm2. So I forget about options available in the larger reMarkable ecosystem :)

**Comment by NiLuJe** [comment]
Might check the uevent feed to see if there's something significant happening on resume, because we *do* listen to that, and it could be used to emit a synthetic "Resume" input event.

**Comment by NiLuJe** [comment]
And/or if systemd itself provides something hookable. (No idea, I avoid systemd like the plague on all my systems).

**Comment by Eeems** [comment]
> uevent

Pretty sure that the uevent module isn't available on the reMarkable 1 or 2

> And/or if systemd itself provides something hookable. (No idea, I avoid systemd like the plague on all my systems).
> 
> (Or, hell, just inject a synthetic input event from there).

Have you looked at my link? That's what I'm suggesting ;) That or you could use the systemd D-Bus API for the hook, which is the recommended way.

**Comment by NiLuJe** [comment]
I meant the actual [Kernel uevent feed](https://github.com/koreader/koreader-base/blob/master/input/libue.h) ;). 

----

Yeah, I did, which is probably why I mentioned it ;p. But I meant that somewhat the other way around: send something from systemd that KOReader kinda already listens to (e.g., input, uevent).

D-Bus is a monstrosity that we'd rather like to avoid, unless necessary (e.g., Kindle, where we have a luxury of native custom wrappers with Lua bindings).

**Comment by snelg** [comment]
Following up, this works on my rm2:

 - Commented out the [`frontend/device/remarkable/device.lua` lines](https://github.com/koreader/koreader/blob/master/frontend/device/remarkable/device.lua#L187-L196) that manually trigger a `UIManager:resume` call after an arbitrary timeout after `systemctl suspend`

 - Created an executable file `/lib/systemd/system-sleep/rm2-wakeup-input.sh` :
```
#!/usr/bin/env bash

if [[ "${1}" == "post" ]]; then
    # simulate the same four events that are generated when the power button is pressed and released
    # (leaving event.time values at 0 for simplicity)

    # type: EV_KEY, code: 116 (power button), value: EV_PRESSED
    echo -n -e '\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x74\x00\x01\x00\x00\x00' > /dev/input/event0

    # type: EV_SYN, code: SYN_REPORT, value: 0
    echo -n -e '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' > /dev/input/event0

    # type: EV_KEY, code: 116 (power button), value: EV_RELEASED
    echo -n -e '\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x74\x00\x00\x00\x00\x00' > /dev/input/event0

    # type: EV_SYN, code: SYN_REPORT, value: 0
    echo -n -e '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' > /dev/input/event0
fi
```

With those changes, koreader cleanly wakes up from sleep.
Xochitl's behavior appears unaffected if I leave that `rm2-wakeup-input.sh` file in `system-sleep`. That's good; I was worried I'd have to hack even more code to add/remove that file when starting/stopping koreader. I don't know what xochitl is doing to know that the device woke up, but I guess it's not simply watching `/dev/input/event0` like koreader does.

So this solution works, but I haven't made a PR because I don't know the preferred way to integrate this into the overall system. My previous PRs have all been changes inside the koreader lua runtime environment. This would need to be part of setup/installation.

**Comment by NiLuJe** [comment]
Yeah, we don't really have a way to *integrate* this kind of things into KOReader.

Which means, the least hacky way I could think about would be:

* Document it in the install instructions
* Update the code, but in a way that make it "work" in case it's not there (e.g., keep the current behavior if the file doesn't exist).

**Comment by NiLuJe** [comment]
And, if you're worried about integration with other things, you can instead generate a completely different keycode and just update the mappings on KOReader's side ;).

**Comment by NiLuJe** [comment]
Fixed via #7345
